src/ConstructiveSolidGeometry/SurfacePrimitives/TorusMantle.jl

```diff
@@ -1,145 +1,269 @@
-    _isapprox_torr_r_tube(p, t.r_torus, t.r_tube, t.z) && _in_φ(p, t.φ) && _in_torr_θ(p, t.r_torus, t.θ, t.z)
+Solve: `solve (sqrt((L1 + λ*D1)^2 + (L2 + λ*D2)^2)-R)^2 + (L3 + λ*D3)^2 = r^2 for λ`
+"""
+function intersection(tm::TorusMantle{T}, l::Line{T}) where {T}
+    obj_l = _transform_into_object_coordinate_system(l, tm) # direction is not normalized
+    
+    L1 = obj_l.origin.x
+    L2 = obj_l.origin.y
+    L3 = obj_l.origin.z
+    D1 = obj_l.direction.x
+    D2 = obj_l.direction.y
+    D3 = obj_l.direction.z
 
-get_φ_limits(t::TorusMantle{T, Nothing}) where {T} = (T(0), T(2π), true)
-get_φ_limits(t::TorusMantle{T, <:AbstractInterval}) where {T} = (t.φ.left, t.φ.right, false)
+    R = tm.r_torus
+    r = tm.r_tube
 
-get_θ_limits(t::TorusMantle{T, <:Any, Nothing}) where {T} = (T(0), T(2π), true)
-get_θ_limits(t::TorusMantle{T, <:Any, <:AbstractInterval}) where {T} = (t.θ.left, t.θ.right, false)
+    A = L1^2 + L2^2 + L3^2 + R^2 - r^2
+    B = 2*(L1*D1 + L2*D2 + L3*D3)
+    C = D1^2 + D2^2 + D3^2
 
+    a = (2*B*C) / C^2
+    b = (2*A*C + B^2 - 4*R^2*(D1^2 + D2^2)) / C^2
+    c = (2*A*B - 8*R^2*(L1*D1 + L2*D2)) / C^2
+    d = (A^2 - 4*R^2*(L1^2 + L2^2)) / C^2
 
-function sample(t::TorusMantle{T}, step::Real) where {T}
-    φMin::T, φMax::T, _ = get_φ_limits(t)
-    θMin::T, θMax::T, _ = get_θ_limits(t)
-    samples = [
-        CylindricalPoint{T}(get_r_at_θ(t,θ),φ,t.r_tube*sin(θ)+t.z)
-        for φ in (t.r_tube == 0 ? φMin : φMin:step/t.r_tube:φMax)
-        for θ in (t.r_tube == 0 ? θMin : θMin:step/t.r_tube:θMax)
-    ]
+	# λ1, λ2, λ3, λ4 = roots_of_4th_order_polynomial(a, b, c, d) # That does not work for all combinations of a, b, c, d...
+	# fallback to Polynomials.jl, which is slower... We should improve `roots_of_4th_order_polynomial`... 
+	λ1, λ2, λ3, λ4 = real.(Polynomials.roots(Polynomial((d, c, b, a, one(T)))))
+	
+	ints1 = obj_l.origin + λ1 * obj_l.direction 
+    ints2 = obj_l.origin + λ2 * obj_l.direction 
+    ints3 = obj_l.origin + λ3 * obj_l.direction 
+    ints4 = obj_l.origin + λ4 * obj_l.direction 
+    return _transform_into_global_coordinate_system(ints1, tm), 
+	_transform_into_global_coordinate_system(ints2, tm),
+	_transform_into_global_coordinate_system(ints3, tm),
+	_transform_into_global_coordinate_system(ints4, tm)
 end
 
-function sample(t::TorusMantle{T}, Nsamps::NTuple{3,Int}) where {T}
-    φMin::T, φMax::T, _ = get_φ_limits(t)
-    θMin::T, θMax::T, _ = get_θ_limits(t)
-    samples = [
-        CylindricalPoint{T}(get_r_at_θ(t,θ),φ,t.r_tube*sin(θ)+t.z)
-        for φ in (Nsamps[2] ≤ 1 ? φMin : range(φMin, φMax, length = Nsamps[2]))
-        for θ in (Nsamps[3] ≤ 1 ? θMin : range(θMin, θMax, length = Nsamps[3]))
-    ]
-end
 
+"""
+    roots_of_4th_order_polynomial(a::T, b::T, c::T, d::T, e::T)
+	
+Calculate the 4 (possible) roots of `x^4 + ax^3 + bx^2 + cx + d = 0`
+"""
+function roots_of_4th_order_polynomial(a::T, b::T, c::T, d::T) where {T}
```

Comment:
Is this used or WIP?

Reply:
WIP. 
The `Polynomials.roots` method is not specialized a 4th order polynomial and causes some allocations... 
But, my implementation fails `roots_of_4th_order_polynomial` fails for certain combination of `a, b, c, d`.